perf: optimize regex cache with LRU eviction and selective pruning

GhostXia opened the pull request:
**Note: This is an initial proposal.**

Based on the **"regex performance bottleneck"** mentioned in #4784, this is my second proposed solution, focusing on optimizing the caching layer:

---

### **Current Behavior**
- Cache is fully cleared on every context switch → **results in 60–80% unnecessary regex recompilations**
- No upper bound on cache size → **risk of uncontrolled memory growth**
- Group chats apply regexes based on `this_chid` instead of the actual speaker → **incorrect regex application**

---

### **What This PR Proposes**
- **Selective pruning:** only clears affected regex categories (e.g., `SCOPED` on character switch) → preserves `GLOBAL` & `PRESET` regexes  
- **LRU eviction (100-entry cap):** ensures predictable and bounded memory usage  
- **Dynamic character loading:** loads regexes based on `characterOverride` parameter → fixes group chat scenarios

GhostXia committed:
Add files via upload

perf: optimize streaming performance with adaptive throttling
- Implement adaptive throttling in StreamingProcessor to reduce UI lag.
- Dynamically adjust DOM update frequency based on render duration.
- Fix freezing issues when using multiple regex scripts.

GhostXia committed:
Add files via upload

feat: Enhance streaming with hybrid adaptive throttling

Implement a hybrid throttling mechanism that combines user-configured FPS
settings with performance-based adaptive throttling for streaming responses.

Changes:
- Modified StreamingProcessor to use dual-layer throttling:
  1. Stopwatch layer: Respects user's streaming_fps setting
  2. Adaptive layer: Dynamically adjusts based on DOM update performance
  
- Updated dynamicThrottleDelay calculation to consider both:
  - Performance-based delay (updateDuration * 2, targeting 33% CPU usage)
  - User-configured minimum delay (1000 / streaming_fps)
  - Takes the larger value to ensure both constraints are satisfied
  
- Initialized dynamicThrottleDelay to 0 for immediate first update

Benefits:
- Respects user's FPS preference as an upper limit
- Automatically reduces FPS on low-performance devices to prevent UI freeze
- Future-proof: When streaming_fps is removed/set to 0, code automatically
  becomes fully adaptive without modification
- Supports high refresh rate displays (144Hz+) when user configures accordingly
- Maintains 1000ms cap to protect low-end devices (minimum 1 FPS)

Technical details:
- Actual delay = min(1000ms, max(adaptiveDelay, userMinDelay))
- Adaptive delay = DOM update duration * 2 (33% CPU target)
- User min delay = 1000 / streaming_fps (0 if not set)

GhostXia committed:
Add files via upload

refactor: Improve streaming throttle with smart defaults and emergency protection

- Initialize dynamicThrottleDelay with user's FPS setting or 50ms default
  (previously 0) for better initial performance
  
- Add emergency protection: enforce 400ms minimum delay when DOM update
  exceeds 200ms to prevent severe lag on low-end devices

Technical:
- Initial delay = streaming_fps ? (1000 / streaming_fps) : 50
- Performance protection = updateDuration > 200 ? 400 : 0
- Final delay = min(1000, max(adaptiveDelay, userMinDelay, performanceProtection))

GhostXia committed:
Add files via upload

feat: optimize regex extension performance

- Implemented caching mechanism for regex scripts to improve message rendering speed.
- Added automatic cache invalidation on script updates.
- Optimized regex compilation to reduce CPU usage.

GhostXia committed:
Add files via upload

GhostXia committed:
Add files via upload

feat: enable automatic streaming throttling on mobile devices

- Add device detection during StreamingProcessor initialization
- Apply a default 50ms throttle delay on mobile devices (iOS/Android) to improve performance
- Keep desktop devices unthrottled by default (0ms)
- User-defined `streaming_fps` settings take precedence over default values

GhostXia committed:
Add files via upload

feat: enable adaptive throttling by default on mobile devices

- Automatically enable adaptive throttling on mobile devices (iOS/Android) to improve performance and prevent stuttering
- Keep adaptive throttling disabled by default on desktop, with optional manual activation via `power_user.adaptive_throttling`

GhostXia committed:
Add files via upload

fix: restore missing changes from commit c87725f

- The modifications from commit c87725f were not properly included
- This commit re-applies the intended updates from c87725f

GhostXia committed:
Update script.js

GhostXia committed:
feat: 添加主应用脚本 `script.js`。

GhostXia committed:
feat: 添加正则表达式脚本引擎，用于管理全局、作用域和预设正则表达式。

GhostXia committed:
feat: 引入正则表达式引擎，用于管理全局、预设和作用域正则表达式脚本，并支持权限和缓存。

GhostXia committed:
feat: 实现正则表达式脚本管理引擎，支持全局、范围和预设脚本类型，并包含权限控制。

GhostXia committed:
Merge branch 'SillyTavern:release' into regex-perf

GhostXia committed:
恢复 script.js 到 staging 分支状态

DeclineThyself commented:
@GhostXia Why did you revert https://github.com/SillyTavern/SillyTavern/pull/4791?

https://github.com/SillyTavern/SillyTavern/pull/4803/files#diff-4a4ec86e4c05b413e9b128d0f702c797e4a2afd8be513301c019824560d0c8e6L2601-L2641

GhostXia commented:
> @GhostXia Why did you revert #4791?
> 
> https://github.com/SillyTavern/SillyTavern/pull/4803/files#diff-4a4ec86e4c05b413e9b128d0f702c797e4a2afd8be513301c019824560d0c8e6L2601-L2641

My repo was not up-to-date, so I was working with an older version of the files. I just pulled the latest staging branch to update and don’t know why the changes got reverted.

GhostXia committed:
Add files via upload

更新到最新的staging分支文件

GhostXia commented:
> @GhostXia Why did you revert #4791?
> 
> https://github.com/SillyTavern/SillyTavern/pull/4803/files#diff-4a4ec86e4c05b413e9b128d0f702c797e4a2afd8be513301c019824560d0c8e6L2601-L2641

`public/script.js` should now be completely identical to the `staging` branch.

Cohee1207 commented:
JFYI

> English is the primary language of communication in this project. Please use only English **when writing commit messages**, PR descriptions, comments and other text. This does not apply to contributions to localization files.

https://github.com/SillyTavern/SillyTavern/blob/release/CONTRIBUTING.md

Cohee1207 commented:
> Group chats apply regexes based on this_chid instead of the actual speaker → incorrect regex application

This is 100% expected and documented. Scoped regex don't work in groups.

Cohee1207 commented on public/scripts/extensions/regex/engine.js:
```diff
@@ -34,14 +34,128 @@ export const SCRIPT_TYPE_UNKNOWN = -1;
+
+    // Dynamic Character Loading for Group Chats
+    if (characterOverride) {
+        const overrideId = characters.findIndex(c => c.name === characterOverride);
```
Looking for a character by name alone is collision-prone.

Cohee1207 commented on public/scripts/extensions/regex/engine.js:
```diff
@@ -34,14 +34,128 @@ export const SCRIPT_TYPE_UNKNOWN = -1;
 const DEFAULT_GET_REGEX_SCRIPTS_OPTIONS = Object.freeze({ allowedOnly: false });
 
+// Cache for regex scripts to avoid rebuilding the array on every call
+let scriptsCacheAll = null;
```
It will be better to group all cache-related fields and functions into a class for better scope isolation.
```js
class RegexCache {
  ...
  get()
  ...
  set()
  ...
  prune()
}
```

(names are not final)

GhostXia committed:
- Replace character lookup by name with avatar to avoid collisions
- Refactor regex cache into RegexCache class for better encapsulation
- Simplify cache operations in runRegexScript (28 lines → 12 lines)

GhostXia committed:
Merge branch 'regex-perf' of https://github.com/GhostXia/SillyTavern into regex-perf

GhostXia commented:
I think I’ve probably completed refactoring cache-related fields and functions into `RegexCache`.

Cohee1207 committed:
Merge branch 'staging' into regex-perf

Cohee1207 commented on public/scripts/extensions/regex/engine.js:
```diff
@@ -34,14 +34,198 @@ export const SCRIPT_TYPE_UNKNOWN = -1;
+    /**
+     * Gets a cached regex entry.
+     * @param {string} regexString The regex string to look up
+     * @returns {{ regex: RegExp, types: Set<number> } | undefined} The cached entry or undefined
```
What's the significance of storing types? A regex string parsing result won't actually change regardless of where it is used.

Cohee1207 commented on public/scripts/extensions/regex/engine.js:
```diff
@@ -34,14 +34,198 @@ export const SCRIPT_TYPE_UNKNOWN = -1;
+ */
+class RegexCache {
+    #cache = new Map();
+    #maxSize = 100;
```
I think this can be safely raised to 500 even.

Cohee1207 commented:
By the way, did you know that global and sticky regex (the ones that have `g`/`y` flags in them) are stateful in JavaScript?
So you need to clone the regex object when returning it from the cache instead of just returning a raw instance.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/test
`const copy = new RegExp(original.source, original.flags);`

GhostXia committed:
Refactor regex caching: fix state leakage and simplify logic

- Fix issue where stateful regexes (global/sticky) shared `lastIndex` state.
- Simplify cache implementation by removing type tracking and pruning.
- Increase cache size limit to 500 for better performance.
- Update regex script execution to use the simplified cache API.

Cohee1207 commented:
So the implication here is that most of the CPU time is spent parsing the regex scripts from string, not actually evaluating them? Can you please provide performance profiles before/after this change that prove that regex parsing really was the bottleneck?

GhostXia commented:
Good News
Base:
<img width="237" height="240" alt="d4e6b981373c7f9056c8f23a5a6cc06b" src="https://github.com/user-attachments/assets/76f232b4-7692-4e2b-b2ab-3d5e0663867c" />
LRU:
<img width="232" height="239" alt="8fd85936f3644aed789d93913c760eef" src="https://github.com/user-attachments/assets/49b494d1-8ee0-450b-8dcb-37e6d2af2a54" />

Bad news — I accidentally synced to the main branch (；´д｀)ゞ

GhostXia commented:
LRU:
<img width="957" height="340" alt="5616eab1e8654e6dca97dab5f4e26417" src="https://github.com/user-attachments/assets/7c67c7b7-b80f-4d1c-9026-339816bc80a2" />
Base:
<img width="939" height="319" alt="b025ba08e847306f6223daa2a5c3ce47" src="https://github.com/user-attachments/assets/5997cc7d-d514-411d-9eb5-3f6ec2835238" />

GhostXia committed:
Merge branch 'staging' into regex-perf
